File: docs/snippets/all/archetypes/audio_clip.py
"""Create and log a multi-channel audio clip."""

import numpy as np

import rerun as rr

rr.init("rerun_example_audio_clip", spawn=True)

sample_rate = 16_000
seconds = 2.0
t = np.arange(int(sample_rate * seconds), dtype=np.float32) / sample_rate

left = 0.4 * np.sin(2.0 * np.pi * 220.0 * t)
right = 0.4 * np.sin(2.0 * np.pi * 330.0 * t)
samples = np.stack([left, right], axis=1).astype(np.float32)

rr.set_time("time", duration=0.0)
rr.log(
    "audio",
    rr.AudioClip(
        samples,
        sample_rate=sample_rate,
        channel_names=["left", "right"],
    ),
)

File: docs/snippets/all/archetypes/audio_clip.rs
//! Create and log a multi-channel audio clip.

use ndarray::{Array2, ShapeBuilder as _};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec =
        rerun::RecordingStreamBuilder::new("rerun_example_audio_clip")
            .spawn()?;

    let sample_rate = 16_000.0;
    let seconds = 2.0;
    let sample_count = (sample_rate * seconds) as usize;

    let samples = Array2::from_shape_fn((sample_count, 2).f(), |(i, channel)| {
        let t = i as f32 / sample_rate as f32;
        let frequency = if channel == 0 { 220.0 } else { 330.0 };
        0.4 * (std::f32::consts::TAU * frequency * t).sin()
    });

    rec.set_time("time", rerun::Duration::from_secs(0.0));
    let samples =
        rerun::datatypes::TensorData::try_from(samples)?.with_dim_names(["sample", "channel"]);

    rec.log(
        "audio",
        &rerun::AudioClip::new(samples, sample_rate).with_channel_names(["left", "right"]),
    )?;

    Ok(())
}
